### PR TITLE
Basic Ranking Capabilities

### PR DESCRIPTION
**Pull request by stopping**
Features added:
- Course and class scores are calculated eagerly and cached
- Courses can be ranked either by quality or difficulty
- Fixed issue where course IDs with underscores couldn't load

Other changes:
- CacheableLazyLoadedPage is now concrete due to difficulties with field
  injection
- DataAggregator intends to perform manipulations on bulk data
  (filtering, ranking, etc)

### CODE REVIEW

File: src/main/java/com/whichclasses/frontend/SearchApiServlet.java [hunk 1]
```diff
@@ -29,8 +35,31 @@ protected void doGet(HttpServletRequest request, HttpServletResponse response)
...
+      break;
+    }
+    
+    String json = "{\"results\":[";
+    List<Course> courses = mDataAggregator.getRankedCourses(department, question);
+    for(Course c : courses) {
+    	String[] split = c.toString().split(" - ", 2);
+    	String number = split[0];
+    	String title = split[1];
+    	// It works.
```
Comment (gunsch) [inline src/main/java/com/whichclasses/frontend/SearchApiServlet.java:58]: Look into GSON. Used a few other places in the code already. Don't write the JSON yourself.

File: src/main/java/com/whichclasses/frontend/SearchApiServlet.java [hunk 2]
```diff
@@ -29,8 +35,31 @@ protected void doGet(HttpServletRequest request, HttpServletResponse response)
...
     response.setContentType("application/json");
-    // TODO: actually implement this
-    response.getWriter().write("{\"results\":[{\"identifier\":{\"department\":\"ECE\",\"course_number\":\"351A\"},\"title\":\"All The Circuits\"}]}");
+    String quality = request.getParameter("quality");
+    String entityType = request.getParameter("entityType");
+    String department = request.getParameter("department");
+    
+    Question question = Question.OVERALL_COURSE_RATING;
+    switch(quality) {
+    case "best":
```
Comment (gunsch) [inline src/main/java/com/whichclasses/frontend/SearchApiServlet.java:44]: Style nit: indent case statements (https://google-styleguide.googlecode.com/svn/trunk/javaguide.html#s4.8.4-switch)

In general, you'll find this project following the Google Java style guide.

File: src/main/java/com/whichclasses/frontend/SearchApiServlet.java [hunk 3]
```diff
@@ -29,8 +35,31 @@ protected void doGet(HttpServletRequest request, HttpServletResponse response)
...
     response.setStatus(HttpServletResponse.SC_OK);
     response.setContentType("application/json");
-    // TODO: actually implement this
-    response.getWriter().write("{\"results\":[{\"identifier\":{\"department\":\"ECE\",\"course_number\":\"351A\"},\"title\":\"All The Circuits\"}]}");
+    String quality = request.getParameter("quality");
+    String entityType = request.getParameter("entityType");
+    String department = request.getParameter("department");
+    
+    Question question = Question.OVERALL_COURSE_RATING;
+    switch(quality) {
```
Comment (gunsch) [inline src/main/java/com/whichclasses/frontend/SearchApiServlet.java:43]: Actually, this whole switch statement should just be a Map<String, Question>. If you don't find the key, return a 400 (bad request).

File: src/main/java/com/whichclasses/model/TceClassModel.java [hunk 4]
```diff
@@ -24,4 +28,45 @@ public TceClassModel(TceClassProto proto) {
   @Override public String toString() {
     return mProto.toString();
   }
+
+  @Override
+  public double getAverageScore(TceRating.Question question) {
+    List<TceRating> ratings = mProto.getRatingList();
+    for(TceRating rating : ratings) {
```
Comment (gunsch) [inline src/main/java/com/whichclasses/model/TceClassModel.java:35]: style nit: space between "for" and "(" (throughout this file)

File: src/main/java/com/whichclasses/scraper/Scraper.java [hunk 5]
```diff
@@ -22,7 +22,7 @@ public Scraper(DeptListPage deptListPage) {
   public void runScrape() {
     // For now: scrape one thing and build models for each.
     Map<String, Department> depts = mDeptList.getChildren();
-    Department firstDepartment = depts.get("ACCT");
+    Department firstDepartment = depts.get("C_SC");
```
Comment (gunsch) [inline src/main/java/com/whichclasses/scraper/Scraper.java:25]: lol

File: src/main/java/com/whichclasses/scraper/Scraper.java [hunk 6]
```diff
@@ -30,8 +30,8 @@ public void runScrape() {
       System.out.println("Got course " + oneCourse);
       Map<String, TceClass> classes = oneCourse.getChildren();
       for (TceClass oneClass : classes.values()) {
-//        System.out.println("Got class " + oneClass);
-//        System.out.println("Class: " + oneClass.getModel().toString());
+        System.out.println("Got class " + oneClass);
+        //System.out.println("Class: " + oneClass.getModel().toString());
```
Comment (gunsch) [inline src/main/java/com/whichclasses/scraper/Scraper.java:34]: please revert this part of the change, or maybe all the changes to this file. not relevant for this pull request

File: src/main/java/com/whichclasses/scraper/page/DeptListPage.java [hunk 7]
```diff
@@ -60,8 +59,24 @@ private DeptListModel getModel() {
    * @return
    */
   private String normalizeDepartmentId(String name) {
-    return name.replaceAll("[\\s_]", "");
+    return name.replaceAll("[\\s]", "");
+    //return name.replaceAll("[\\s_]", "");
```
Comment (gunsch) [inline src/main/java/com/whichclasses/scraper/page/DeptListPage.java:63]: ?

Comment (stopping) [inline src/main/java/com/whichclasses/scraper/page/DeptListPage.java:63]: TCE website is stupid, for some reason it uses underscores in the 'crssub' query field for the department list links, but doesn't use them in the course list links.

E.g.:
Department List:
href="GenerateReport.aspx?Report=DEPTCOURSE&amp;&amp;<b>crssub=C_SC</b>&amp;&amp;deptnm=Computer+Science+(C+SC)"

Course:
href="GenerateReport.aspx?Report=DEPTONECLASS&amp;&amp;crstitle=Introduction+to+Computer+Science&amp;<b>crssub=CSC</b>&amp;crsnum=127A"

**Comment by gunsch** [comment]
This is a really massive PR and hasn't had any activity. I'm hoping to start working on the site again this week and might end up clobbering some of this. I'd suggest breaking it into smaller PRs to be able to more easily make progress moving forward.

**Pull request closed without merging**